state-chain/pallets/cf-ingress-egress/src/lib.rs

```diff
@@ -1200,36 +1196,49 @@ impl<T: Config<I>, I: 'static> EgressApi<T::TargetChain> for Pallet<T, I> {
 			*id
 		});
 		let egress_id = (<T as Config<I>>::TargetChain::get(), egress_counter);
-		match maybe_ccm_with_gas_budget {
+		let egress_fee = match maybe_ccm_with_gas_budget {
 			Some((
 				CcmDepositMetadata { source_chain, source_address, channel_metadata },
 				gas_budget,
-			)) => ScheduledEgressCcm::<T, I>::append(CrossChainMessage {
-				egress_id,
-				asset,
-				amount,
-				destination_address: destination_address.clone(),
-				message: channel_metadata.message,
-				cf_parameters: channel_metadata.cf_parameters,
-				source_chain,
-				source_address,
-				gas_budget,
-			}),
-			None => ScheduledEgressFetchOrTransfer::<T, I>::append(FetchOrTransfer::<
-				T::TargetChain,
-			>::Transfer {
-				asset,
-				destination_address: destination_address.clone(),
-				amount,
-				egress_id,
-			}),
-		}
+			)) => {
+				ScheduledEgressCcm::<T, I>::append(CrossChainMessage {
+					egress_id,
+					asset,
+					amount,
+					destination_address: destination_address.clone(),
+					message: channel_metadata.message,
+					cf_parameters: channel_metadata.cf_parameters,
+					source_chain,
+					source_address,
+					gas_budget,
+				});
+				gas_budget
+			},
+			None => {
+				let (amount_after_fees, egress_fee) =
+					Self::withhold_transaction_fee(IngressOrEgress::Egress, asset, amount);
+				ScheduledEgressFetchOrTransfer::<T, I>::append({
+					FetchOrTransfer::<T::TargetChain>::Transfer {
+						asset,
+						destination_address: destination_address.clone(),
+						amount: amount_after_fees,
+						egress_id,
+					}
+				});
+				egress_fee
+			},
+		};
+
+		DepositBalances::<T, I>::mutate(asset, |tracker| {
+			tracker.register_transfer(amount);
```

Comment:
Both `tracker.register_transfer` and `tracker.mark_as_fetched` used to be invoked in `fn 
do_egress_scheduled_fetch_transfer`.

One of them is moved into `fn schedule_egress` as the original (gross?) amount is not saved after taking the fee.

Can the `tracker.mark_as_fetched` invocation be moved to the `fn schedule_egress` too?
IMO, it would make easier to reason about storage changes this way.

Reply:
For `mark_as_fetched`, I think the equivalent place would be `process_single_deposit` (where we append the fetch to `ScheduledEgressFetchOrTransfer`), but it is also where `deposits.register_deposit` is called (which adds the amount as "unfetched"). To be honest, I'm not 100% sure what the distinction between fetched/unfetched is (seems like all unfetched funds become "fetched" at the end of the block, so maybe the intention is to add a 1 block delay until we can use them?).

Reply:
Yeah I'm not sure what Dan intended by fetched/unfetched in DepositBalances here. But given it's not used anywhere else except tracking I don't think it matters a whole lot *when* things occur, so long as the amounts are correct. I think leaving the fetched where the tx is kicked off (where it is now).

Reply:
I actually think there's a bug here after looking at this a bit deeper now too. In `process_single_deposit` we add the `net_deposit_amount` to the DepositBalances. We pass through the `deposit_amount` to the fetch transaction. When we process that fetch in `do_egress_scheduled_fetch_transfer` we do:

```rust
FetchOrTransfer::<T::TargetChain>::Fetch {
					asset,
					deposit_address,
					deposit_fetch_id,
					amount,
				} => {
					fetch_params.push(FetchAssetParams {
						deposit_fetch_id: deposit_fetch_id.expect("Checked in extract_if"),
						asset,
					});
					addresses.push(deposit_address.clone());
					DepositBalances::<T, I>::mutate(asset, |tracker| {
						tracker.mark_as_fetched(amount);
					});
```

We've now fetched more than we had in unfetched.

Reply:
I changed it so that the unfetched adds the deposit_amount rather than the amount_after_fees